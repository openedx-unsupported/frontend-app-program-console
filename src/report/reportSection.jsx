import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { sendTrackEvent } from '@edx/frontend-platform/analytics';
import { injectIntl } from '@edx/frontend-platform/i18n';
import { Collapsible, Alert } from '@openedx/paragon';

import { fetchReports } from './actions';
import { reportSelector, storeName } from './selectors';
import reducer from './reducers';
import saga from './sagas';
import { configureApiService } from './service';

export class ReportSection extends React.Component {
  componentDidMount() {
    if (this.props.programKey) {
      this.props.fetchReports(this.props.programKey);
    }
  }

  getDisclaimer = () => (
    <Alert
      variant="info"
      dismissible={false}
      show
    >
      The data contained in these reports reflect enrollments only and are not intended to be used for
      financial reporting or reconciliation.
    </Alert>
  );

  render() {
    if (this.props.reportData[this.props.programKey]
      && this.props.reportData[this.props.programKey].length > 0) {
      return (
        <Collapsible
          className="shadow"
          title="Download Reports"
          defaultOpen={this.props.isFirstSection}
        >
          <div className="container">
            {this.getDisclaimer()}
            {this.props.reportData[this.props.programKey].map(report => (
              <div key={report.name}>
                <a
                  id="console-report"
                  href={report.downloadUrl}
                  onClick={() => sendTrackEvent(
                    'edx.program.console.report.download',
                    {
                      programKey: this.props.programKey,
                      reportName: report.name,
                    },
                  )}
                >
                  {report.name}
                </a>
              </div>
            ))}
          </div>
        </Collapsible>
      );
    }
    return null;
  }
}

ReportSection.propTypes = {
  reportData: PropTypes.shape().isRequired,
  fetchReports: PropTypes.func.isRequired,
  programKey: PropTypes.string.isRequired,
  isFirstSection: PropTypes.bool.isRequired,
};

export default connect(reportSelector, { fetchReports })(injectIntl(ReportSection));

export {
  reducer,
  saga,
  configureApiService,
  storeName,
};
