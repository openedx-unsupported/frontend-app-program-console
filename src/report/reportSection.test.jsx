import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import * as analytics from '@edx/frontend-platform/analytics';
import { ReportSection } from './reportSection';

jest.mock('@edx/frontend-platform/analytics', () => ({
  sendTrackEvent: jest.fn(),
}));

describe('ReportSection component', () => {
  it('renders with the most basic props passed to it', () => {
    const { container: tree } = render(
      <ReportSection
        reportData={{}}
        fetchReports={() => { }}
        programKey=""
        isFirstSection
      />,
    );
    expect(tree).toMatchSnapshot();
  });

  describe('correctly passes isFirstSection prop to collapsible component as defaultOpen', () => {
    const reportData = {
      'program-key': [
        {
          name: 'one-report',
          downloadUrl: 'example.com/one-report',
        },
      ],
    };
    [true, false].forEach((isFirstSectionValue) => {
      it(`${isFirstSectionValue}`, () => {
        const reportSectionComponent = (
          <ReportSection
            reportData={reportData}
            fetchReports={() => { }}
            programKey="program-key"
            isFirstSection={isFirstSectionValue}
          />
        );
        render(reportSectionComponent);
        const collapsible = screen.getByRole('button', { name: 'Download Reports' });
        const isExpanded = collapsible.getAttribute('aria-expanded') === 'true';
        expect(isExpanded).toBe(isFirstSectionValue);
      });
    });
  });

  it('renders contents when program data is passed in', () => {
    const reportData = {
      'program-key': [
        {
          name: 'first-report',
          downloadUrl: 'example.com/first-report',
        },
        {
          name: 'second-report',
          downloadUrl: 'example.com/second-report',
        },
      ],
    };

    const reportSectionComponent = (
      <ReportSection
        reportData={reportData}
        fetchReports={() => { }}
        programKey="program-key"
        isFirstSection
      />
    );

    const { container, debug } = render(reportSectionComponent);

    debug();

    expect(container).toMatchSnapshot();
    const collapsible = container.querySelector('.pgn_collapsible');
    expect(collapsible).toHaveClass('shadow');
    expect(collapsible).toHaveTextContent('Download Reports');
    expect(collapsible.querySelector('.container')).toBeInTheDocument();

    const links = screen.getAllByRole('link');
    expect(links).toHaveLength(2);
    links.forEach((link, index) => {
      expect(link).toHaveAttribute('href', reportData['program-key'][index].downloadUrl);
      expect(link).toHaveTextContent(reportData['program-key'][index].name);
    });

    const statusAlert = screen.getByRole('alert');
    expect(statusAlert).toHaveClass('alert-info');
    expect(statusAlert).toHaveTextContent('The data contained in these reports reflect enrollments only and are not intended to be used for financial reporting or reconciliation.');
    expect(statusAlert).toHaveClass('show');
    expect(statusAlert).not.toHaveClass('dismissible');
  });

  it('doesn\'t render contents when no program data is passed in', () => {
    const reportSectionComponent = (
      <ReportSection
        reportData={{}}
        fetchReports={() => { }}
        programKey="program-key"
        isFirstSection
      />
    );

    render(reportSectionComponent);
    const collapsible = screen.queryByRole('button', { name: 'Download Reports' });
    expect(collapsible).not.toBeInTheDocument();
  });

  it('doesn\'t render contents when empty program key is passed in', () => {
    const reportSectionComponent = (
      <ReportSection
        reportData={{}}
        fetchReports={() => { }}
        programKey=""
        isFirstSection
      />
    );

    render(reportSectionComponent);
    const collapsible = screen.queryByRole('button', { name: 'Download Reports' });
    expect(collapsible).not.toBeInTheDocument();
  });

  it('calls the fetchReports function on load if program key prop is passed in', () => {
    const mock = jest.fn();

    render(<ReportSection
      reportData={{}}
      fetchReports={mock}
      programKey="program-key"
      isFirstSection
    />);

    expect(mock).toHaveBeenCalled();
  });

  it('doesn\'t call the fetchReports function on load if program key prop is not passed in', () => {
    const mock = jest.fn();

    render(<ReportSection
      reportData={{}}
      fetchReports={mock}
      programKey=""
      isFirstSection
    />);

    expect(mock).not.toHaveBeenCalled();
  });

  it('calls sendTrackEvent when report download link is clicked', () => {
    const reportData = {
      'program-key': [
        {
          name: 'first-report',
          downloadUrl: 'example.com/first-report',
        },
      ],
    };

    const reportSectionComponent = (
      <ReportSection
        reportData={reportData}
        fetchReports={() => { }}
        programKey="program-key"
        isFirstSection
      />
    );

    const { container } = render(reportSectionComponent);
    const link = container.querySelector('#console-report');
    fireEvent.click(link);
    expect(analytics.sendTrackEvent).toHaveBeenCalled();
  });
});
