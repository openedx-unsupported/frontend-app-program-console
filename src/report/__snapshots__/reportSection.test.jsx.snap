// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportSection component renders contents when program data is passed in 1`] = `
<div>
  <div
    class="pgn_collapsible shadow collapsible-card is-open"
  >
    <div
      aria-expanded="true"
      class="collapsible-trigger"
      role="button"
      tabindex="0"
    >
      <span>
        Download Reports
      </span>
      <span
        class="collapsible-icon"
      >
        <span
          class="pgn__icon"
        >
          <svg
            aria-hidden="true"
            fill="none"
            focusable="false"
            height="24"
            role="img"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m12 8-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14l-6-6Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </span>
    </div>
    <div
      class="pgn-transition-replace-group position-relative"
    >
      <div
        style="padding: .1px 0px;"
      >
        <div
          class="collapsible-body"
        >
          <div
            class="container"
          >
            <div
              class="fade alert-content alert alert-info show"
              role="alert"
            >
              <div
                class="pgn__alert-message-wrapper"
              >
                <div
                  class="alert-message-content"
                >
                  The data contained in these reports reflect enrollments only and are not intended to be used for financial reporting or reconciliation.
                </div>
              </div>
            </div>
            <div>
              <a
                href="example.com/first-report"
                id="console-report"
              >
                first-report
              </a>
            </div>
            <div>
              <a
                href="example.com/second-report"
                id="console-report"
              >
                second-report
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ReportSection component renders with the most basic props passed to it 1`] = `<div />`;
